dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -847,35 +1006,99 @@ PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* array_ptr, const ssize
 
 template <class BinaryOp>
 std::span<const ssize_t> PartialReduceNode<BinaryOp>::axes() const {
-    // TODO: support multiple axes
+    // todo: support multiple axes
     return std::span(axes_.get(), 1);
 }
 
 template <class BinaryOp>
 double const* PartialReduceNode<BinaryOp>::buff(const State& state) const {
-    return data_ptr<ArrayNodeStateData>(state)->buff();
+    return data_ptr<PartialReduceNodeData<op>>(state)->buff();
 }
 
 template <class BinaryOp>
 void PartialReduceNode<BinaryOp>::commit(State& state) const {
-    data_ptr<ArrayNodeStateData>(state)->commit();
+    data_ptr<PartialReduceNodeData<op>>(state)->commit();
 }
 
 template <class BinaryOp>
 std::span<const Update> PartialReduceNode<BinaryOp>::diff(const State& state) const {
-    return data_ptr<ArrayNodeStateData>(state)->diff();
+    return data_ptr<PartialReduceNodeData<op>>(state)->diff();
 }
 
 template <class BinaryOp>
 void PartialReduceNode<BinaryOp>::initialize_state(State& state) const {
-    std::vector<double> values(size(state));
+    // the type we use for the reduction depends on the node type
+    using accumulator_type = std::conditional<std::same_as<BinaryOp, std::multiplies<double>>,
+                                              RunningProduct, double>::type;
+
+    std::vector<accumulator_type> values(size(state));
+
+    const ssize_t axis = axes_[0];
+
+    for (ssize_t index = 0, stop = size(state); index < stop; ++index) {
+        assert(index < static_cast<ssize_t>(values.size()));
+
+        /// We wish to fill `index` of the partial reduction, given the state of the parent.
```

Comment:
This is just the old `reduce()` function moved inside the loop. Because we do some templated type fiddling. But the logic is all the same with the exception of the `constexpr if` later on.

Reply:
I guess the `reduce()` function wasn't used anywhere but `initialize_state()`?

Reply:
Yeah exactly.